core/basic-authorship/src/basic_authorship.rs

```diff
@@ -42,6 +43,9 @@ use substrate_telemetry::{telemetry, CONSENSUS_INFO};
 pub trait BlockBuilder<Block: BlockT> {
 	/// Push an extrinsic onto the block. Fails if the extrinsic is invalid.
 	fn push_extrinsic(&mut self, extrinsic: <Block as BlockT>::Extrinsic) -> Result<(), error::Error>;
+	/// Push an inherent digest onto the block.
+	fn set_inherent_digest(&mut self, inherent_digest: DigestFor<Block>);
```

Comment:
the API here implies that `push_extrinsic` can be done interleaved with `set_inherent_digest`. But my impression is that it's really only something that's done at the beginning.

Reply:
Shouldn't there be a definition that enforces this?
We want inherent digests to show up at the beginning of the header, but if you'd push an extrinsic that produced another digest first, we would get a different result.

Reply:
That is indeed a bug.  Probably the best solution is to pass the inherent digests to the constructor.